packages/ngtools/webpack/src/transformers/find_image_domains.ts

```diff
@@ -0,0 +1,138 @@
+      };
+
+      const findDomainString = (node: ts.Node) => {
+        if (
+          ts.isStringLiteral(node) ||
+          ts.isTemplateHead(node) ||
+          ts.isTemplateMiddle(node) ||
+          ts.isTemplateTail(node)
+        ) {
+          const domain = node.text.match(URL_REGEX);
+          if (domain && domain[0]) {
+            imageDomains.add(domain[0]);
+
+            return node;
+          }
+        }
+        ts.visitEachChild(node, findDomainString, context);
+
+        return node;
+      };
+
+      function isImageProviderKey(property: ts.ObjectLiteralElementLike): boolean {
+        return (
+          ts.isPropertyAssignment(property) &&
+          property.name.getText() === 'provide' &&
+          property.initializer.getText() === 'IMAGE_LOADER'
+        );
+      }
+
+      function isImageProviderValue(property: ts.ObjectLiteralElementLike): boolean {
+        return ts.isPropertyAssignment(property) && property.name.getText() === 'useValue';
+      }
+
+      function checkForDomain(node: ts.ObjectLiteralExpression) {
+        if (node.properties.find(isImageProviderKey)) {
+          const value = node.properties.find(isImageProviderValue);
+          if (value && ts.isPropertyAssignment(value)) {
+            if (
+              ts.isArrowFunction(value.initializer) ||
+              ts.isFunctionExpression(value.initializer)
+            ) {
+              ts.visitEachChild(node, findDomainString, context);
+            }
+          }
+        }
+      }
+
+      function findImageLoaders(node: ts.Node) {
+        if (ts.isCallExpression(node)) {
+          if (isBuiltinImageLoader(node)) {
+            const firstArg = node.arguments[0];
+            if (ts.isStringLiteralLike(firstArg)) {
+              imageDomains.add(firstArg.text);
+            }
+          }
+        } else if (ts.isObjectLiteralExpression(node)) {
+          checkForDomain(node);
+        }
+
+        return node;
+      }
+
+      function findPropertyAssignment(node: ts.Node) {
+        if (ts.isPropertyAssignment(node)) {
+          if (ts.isIdentifier(node.name) && node.name.escapedText === 'providers') {
+            ts.visitEachChild(node.initializer, findImageLoaders, context);
+          }
+        }
+
+        return node;
+      }
+
+      function findPropertyDeclaration(node: ts.Node) {
+        if (
+          ts.isPropertyDeclaration(node) &&
+          ts.isIdentifier(node.name) &&
+          node.name.escapedText === 'ɵinj'
+        ) {
+          if (node.initializer) {
+            if (ts.isCallExpression(node.initializer)) {
+              ts.visitEachChild(node.initializer.arguments[0], findPropertyAssignment, context);
+            }
+          }
+        }
+
+        return node;
+      }
+
+      // Continue traversal if node is ClassDeclaration and has name "AppModule"
+      function findClassDeclaration(node: ts.Node) {
+        if (ts.isClassDeclaration(node) && node.name?.escapedText === 'AppModule') {
```

Comment:
Ah, thanks for pointing this out. I tried to keep this as specific as possible to limit traversal (for performance), but I guess here it just needs to be all ClassDeclaration nodes. Since most files won't have _that_ many top-level class declarations, and the traversal should fail at the next step for any non-appmodule classes, it shouldn't cause too much additional work.